internal/oauth2/lazy_token.go

```diff
@@ -0,0 +1,123 @@
+package oauth2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+)
+
+// Tokener is an interface that defines the Token method.
+type Tokener interface {
+	Token(context.Context) string
+}
+
+// LazyToken is a struct that handles token acquisition and caching.
+type LazyToken struct {
+	url          string
+	clientId     string
+	clientSecret string
+	token        string
+	offlineToken string
+	expiration   time.Time
+	mutex        sync.Mutex
+}
+
+type tokenResponse struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int    `json:"expires_in"`
+	TokenType   string `json:"token_type"`
+}
+
+// NewLazyToken creates a new LazyToken instance.
+func NewLazyToken(url, clientId, clientSecret string) *LazyToken {
+	return &LazyToken{
+		url:          url,
+		clientId:     clientId,
+		clientSecret: clientSecret,
+	}
+}
+
+// Token returns a valid token, acquiring a new one if necessary.
+func (lt *LazyToken) Token(ctx context.Context) string {
+	//echoCtx, ok := ctx.(echo.Context)
+	//	if !ok {
+	//	// Handle the case where the context is not an echo.Context
+	//	return ""
+	//	}
+	lt.mutex.Lock()
+	defer lt.mutex.Unlock()
+
+	if lt.token == "" || lt.expiration.Sub(time.Now()) < time.Minute {
```

Comment:
The `lt.expiration` can be `nil` and it is not being checked.

Before you add `nil` check, here is a trick to consider - in the initializer function, set it to current time. Thus it will always be set to something and you can leave this code as is.